### ISSUE TITLE
Evaluate replacing doxygen with kernel-doc

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#858]
**_Reported by Inaky Perez-Gonzalez:_**

doxygen is yet another tool to use; it works but we have some restrictions by using it.

Evaluate replacing it with kernel-doc, which works as a sphinx plugin and would:

- remove extra dependencies / complexity
- allow us to link inside function docs to rest contents

(Imported from Jira ZEP-952)

**Comment by nashif** [comment]
**_by Inaky Perez-Gonzalez:_**

As a first step, add kernel-doc generation capabilities to the Zephyr tree so we can start testing converting some files.

**Comment by nashif** [comment]
**_by Inaky Perez-Gonzalez:_**

start converting the include/bluetooth.h file, as that is a very complete file in terms of features used/needed.

**Comment by nashif** [comment]
**_by David Kinder:_**

Summary:  Replacing our existing API documentation currently using doxygen/breathe with the linux kernel developers kernel-doc tools (as explained here: https://www.kernel.org/doc/html/latest/doc-guide/index.html) would be a major effort with little benefit over our existing tools and processes.  Conclusion is to continue with existing doxygen/breathe tools.

We will though, improve where documentation exists in the tree (from the developer's point of view) by allowing docs to be placed closer to the source code rather than only in the /doc tree.  This should encourage developers to provide documentation with their subsystem and component source code.

More details:

*  the kernel-doc comment conventions are different than doxygen conventions so we'd have to manually rewrite EVERY doxygen comment into the kernel-doc format
*  kernel-doc does not define "grouping" commands that are used extensively in our doxygen-based API docs - these grouping comments define a collection of API reference material and can come from different source files in the doc tree and multiple groups can be defined within a single source file.
* kernel[-doc does not support cross-referencing within the comments to other entities (doxygen does)
* I took a look at one of the “poster child” projects for linux kernel documentation, the media subsystem.  If you take a look around this site, https://linuxtv.org/downloads/v4l-dvb-apis/index.html you get an idea of how a subsystem can  be documented using restructuredText and Sphinx.  But when you look closer at the documentation, you’ll notice much of it is NOT generated from comments in the code but rather are stand-alone restructuredText documents.  

Bottom line, if we had no existing system in place, kernel-docs could be a simple system to use, but moving to it from our existing system doesn't provide a cost-effective benefit.

**Comment by nashif** [comment]
**_by David Kinder:_**

Recommend closing as "will not do"

**Comment by nashif** [comment]
**_by David Kinder:_**

Evaluation done -- recommendation is not to switch to kernel-doc tools
